src/map/skill.c

```diff
@@ -24724,14 +24760,231 @@ static bool skill_read_skilldb(const char *filename)
+		return;
+	}
+
+	sk->skill_id = skill_id;
+}
+
+/**
+ * Reads a AutoSpell skill's usable levels (SkillLevel) when reading the autospell DB.
+ *
+ * @param conf The libconfig settings block which contains the skill's data.
+ * @param sk The autospell_skill struct where the level should be set.
+ */
+static void skill_read_autospell_skill_level(struct config_setting_t *conf, struct s_autospell_db *sk)
+{
+	nullpo_retv(conf);
+	nullpo_retv(sk);
+
+	skill->level_set_value(sk->skill_lv, 0);
+
+	struct config_setting_t *t = libconfig->setting_get_member(conf, "SkillLevel");
+
+	if (t != NULL && config_setting_is_group(t)) {
+		for (int i = 0; i < MAX_SKILL_LEVEL; i++) {
+			char lv[6]; // Big enough to contain "Lv999" in case of custom MAX_SKILL_LEVEL.
+			snprintf(lv, sizeof(lv), "Lv%d", i + 1);
+
+			int level;
+			if (libconfig->setting_lookup_int(t, lv, &level) == CONFIG_TRUE) {
+				if (level >= 0 && level <= MAX_SKILL_LEVEL)
+					sk->skill_lv[i] = level;
+				else
+					ShowWarning("%s: Invalid SkillLevel %d specified in level %d for skill ID %d in %s! Minimum is 0, maximum is %d. Defaulting to 0...\n",
+						    __func__, level, i + 1, sk->skill_id, conf->file, MAX_SKILL_LEVEL);
+			}
+		}
+
+		return;
+	}
+
+	int level;
+	if (libconfig->setting_lookup_int(conf, "SkillLevel", &level) == CONFIG_TRUE) {
+		if (level >= 0 && level <= MAX_SKILL_LEVEL)
+			skill->level_set_value(sk->skill_lv, level);
+		else
+			ShowWarning("%s: Invalid SkillLevel %d specified for skill ID %d in %s! Minimum is 0, maximum is %d. Defaulting to 0...\n",
+				    __func__, level, sk->skill_id, conf->file, MAX_SKILL_LEVEL);
+	}
+}
+
+/**
+ * Reads additional field settings via plugins when parsing autospell_db.conf
+ * @param   conf    struct, pointer to the entry configuration
+ * @param   sk      struct, struct, pointer to s_autospell_db
+ */
+static void skill_read_autospell_additional_fields(struct config_setting_t *conf, struct s_autospell_db *sk)
+{
+	// Does nothing like a boss. *cough* plugins *cough*
+}
+
+/**
+ * Compares two autospell db entries and sort it as the database expects:
+ * 1. autospell_level = 0 (blank entries) goes to the end
+ * 2. entries are sorted by autospell_level in ascending order
+ *
+ * @param entry1 first entry to compare
+ * @param entry2 second entry to compare
+ * @returns
+ * - < 0 if entry2 should go before entry1 ;
+ * - 0 if entry1 and entry2 are equivalent in sorting ;
+ * - > 0 if entry1 should go before entry2
+ */
+static int skill_autospell_db_entry_compare(const void *entry1, const void *entry2)
+{
+	nullpo_ret(entry1);
+	nullpo_ret(entry2);
+
+	struct s_autospell_db *entry1_ = (struct s_autospell_db *) entry1;
+	struct s_autospell_db *entry2_ = (struct s_autospell_db *) entry2;
+
+	// autospell_level == 0 is always at the end of the list, because they are unused entries
+	if (entry1_->autospell_level == 0)
```

Comment:
This may lead to undefined behavior when called with a value of `0` in both entry1 and entry2 (should return `0`, but it will return `1` here).

Reply:
Fixed. thank you!!